RFD 28: Cluster configuration related resources

andrejtokarcik opened the pull request.

andrejtokarcik committed:
RFD 28: Cluster configuration related resources

klizhentas commented on rfd/0028-cluster-config-resources.md:
```diff
@@ -0,0 +1,176 @@
+RBAC verbs, interpreted as implying additional constraints. Such new verbs
+would be evaluated together with other RBAC verbs in the auth API handlers.
+
+In particular, this is needed for `SessionRecordingConfig` which has two
```
May be call it session config in case we add any other session related parameters?

andrejtokarcik replied:
I considered it but it seemed too vague and potentially confusing given that the term "session" itself can be easily (mis)interpreted in many different ways.

For instance, would the field `SessionControlTimeout` be expected to be grouped under `SessionConfig` or `ClusterNetworkingConfig` (together with the other timeout values)?

klizhentas replied:
ok fair

klizhentas commented on rfd/0028-cluster-config-resources.md:
```diff
@@ -0,0 +1,176 @@
+`ClusterAuthPreference` and `PAMConfig` should also adapted to facilitate
+the dynamic configuration workflow.
+
+### Working with a whole cluster configuration
```
Overall looks good to me. Can you add a couple of practical examples using tctl and resources that I've seen customers requesting:

* Set up session recording node-sync
* Configuring PAM custom variables
* Setting auth preference to OIDC for Auth0
* Setting U2f second factor and CA attestation
* Setting global session U2F as required

andrejtokarcik replied:
Added some simple examples, let me know if this level of detail is sufficient.

andrejtokarcik committed:
Update 0028-cluster-config-resources.md

awly commented on rfd/0028-cluster-config-resources.md:
```diff
@@ -0,0 +1,178 @@
+   or `node-sync`.
+1. The value of `proxy_checks_host_keys` cannot be changed.
+
+This would be achieved by a new RBAC verb `update.no-proxy`. A user could then
+be assigned a role featuring `update.on-proxy` instead of the regular, more
```
nit: let's make these descriptive: `update.proxy_recording_mode` and `update.no_proxy_recording_mode`

andrejtokarcik replied:
There was a typo, I meant to introduce only `update.no-proxy` -- now renamed to `update.no_proxy_recording`.

awly commented on rfd/0028-cluster-config-resources.md:
```diff
@@ -0,0 +1,178 @@
+### Additional dynamically configurable resources
+
+In addition to the resources derived from `ClusterConfig`, the resources
+`ClusterAuthPreference` and `PAMConfig` should also adapted to facilitate
```
```suggestion
`ClusterAuthPreference` and `PAMConfig` should also be adapted to facilitate
```

andrejtokarcik commented on rfd/0028-cluster-config-resources.md:
```diff
@@ -0,0 +1,178 @@
+// DisconnectExpiredCert & LocalAuth fields are moved into ClusterAuthPreference.
+message ClusterAuthPreference {
+    ...
+    bool AllowLocalAuth;
```
`AuthPreference.Type` is just a default authentication type. It does not prevent from using the other types.

andrejtokarcik committed:
Apply suggestions from code review

Co-authored-by: Andrew Lytvynov <andrew@goteleport.com>

andrejtokarcik committed:
Rename 'update.no-proxy' to 'update.no_proxy_recording'

andrejtokarcik committed:
Rename the session recording field to 'mode'

andrejtokarcik committed:
Add examples

alex-kovoy commented on rfd/0028-cluster-config-resources.md:
````diff
@@ -0,0 +1,298 @@
+default set to `node`, here it is already set to `off`.
+
+```
+$ tctl get session_recording_config | tee reccfg.yaml
````
```
$ tctl get cluster_config
$ tctl get cluster_config session_recording
$ tctl get cluster_config pam
$ tctl get cluster_config auth_preference
```

Looks nicer

klizhentas replied:
I don't think `cluster_config` is necessary here

alex-kovoy replied:
That was if we treat `cluster_config` as a parent element but if each is a separate resource then yes it's not necessary.

Btw, I would also consider using a path convention to name these resources: 

```
kind: cluster_config
version: v2
spec:

kind: cluster_config/pam
version: v2
spec:

kind: cluster_config/auth_preferences
version: v2
spec:
```

Having this hierarchy potentially can make RBAC simpler too.  

```
- resources:
      - cluster_config/*
      verbs:
      - list
      - create
      - read
      - update
      - delete
```      

I understand that it might not be possible, sharing it just as an idea.

knisbet reviewed (commented):
So one part I'm curious about is if we want to restrict certain types of changes on the cloud, how that interacts here. For example we might want to tune keep-alives on the cloud, or prevent a customer from breaking the configuration by setting timers longer than the cloud load balancers allow. (Although longer term we might want to try and do a dynamic detection algorithm to detect timeouts on links).

It might also be feasible that we want to limit other types of configurations at some point, for example in a free plan we might limit or disable session recordings as an example.

klizhentas commented:
@knisbet had similar thoughts but from the side of u2f logins. We don't want to allow admins to turn off second factor for the cloud - too risky. How do we achieve that?

andrejtokarcik commented on rfd/0028-cluster-config-resources.md:
```diff
@@ -0,0 +1,298 @@
+### Working with a whole cluster configuration
+
+`KindClusterConfig` should be retained but reinterpreted as a helper meta-kind
+similar to `KindConnectors`. It would allow aggregating all the cluster config
```
I didn't really consider subkinds. I'll have a look and eventually rewrite the proposal. They should allow keeping the whole cluster config in a single resource (auth preference & pam would be then preferable to merge into that one?).

andrejtokarcik replied:
OK, I don't think subkinds model "sub resources" as considered here very well. Subkinds are made to divide a (large) set of `<kind>/<name>` resources that have largely the same structure, into `<kind>/<subkind>/<name>` subsets.

In contrast, here we deal with a couple of resources each of which has its own distinct structure. Moreover each of the config resources is a singleton and so has a fixed pre-determined name that shouldn't be really needed to refer to.

klizhentas replied:
yes, we should split cluster config monolith in parts, and subkind is really to differentiate the different similar kinds of resources, not the parts of the same resource.

andrejtokarcik commented on rfd/0028-cluster-config-resources.md:
```diff
@@ -0,0 +1,298 @@
+  `proxy`, `proxy-sync`;
+* a boolean `proxy_checks_host_keys`.
+
+Some users should be subject to the following restrictions:
```
Well, one use case is precisely limiting the `session_recording` values & possibly also disabling the `second_factor` option, mentioned below.

Regarding `auth_connectors` RBAC, I don't think the situation is that simple. What seems to occur is that both connector-specific *and* `auth_connectors`-wide privileges need to be granted, so for a user to be able to update Github connectors they'd need something like:
```
  allow:
    rules:
    - resources: [github, auth_connector]
      verbs: [read, create, update]
```

klizhentas replied:
@alex-kovoy that's a valid concern and we considered new resources and verbs alternatives. We do need, however a fine-grained control over multiple independent parts of teleport, and having a monolith resource did not scale. So we are pursuing a strategy of splitting the resources at the extra cost of making it a bit harder to use, yes.

alex-kovoy replied:
@andrejtokarcik  I can confirm that `auth_connector` resource is a default way of restricting access to auth connectors. Not their individual parts.

@klizhentas splitting the resource is one thing but polluting the RBAC is another. Do we need it specifically for the cloud use cases? If so, it could be driven by the feature flags to prevent changes of specific config resources from API.

andrejtokarcik replied:
I don't understand this concern with "RBAC polluting". We can provide more comprehensive resource-like subjects but allowing more fine-grained definitions seems much more in line with least privilege policies that we should strive to support.

alex-kovoy replied:
@andrejtokarcik my concern is that we would need to ship a book to teach people how to use RBAC and go around all types of resources that are there. This is why I would stick with a simple `cluster_config` RBAC subject and do the cloud specific constrains in the code for as long as it's manageable.

andrejtokarcik commented:
@knisbet @klizhentas Well, the general approach proposed here is to introduce custom RBAC verbs for such purposes. (The only example considered right now is `update.no_proxy_recording` for the `session_recording_config` resource.)

klizhentas commented:
@andrejtokarcik @knisbet  I think what we are looking for is a constraints resource.

Imagine you have a following resource:

```yaml
kind: cluster_auth_preference
version: v2
spec:
  type: local
  second_factor: u2f
  u2f:
    app_id: https://proxy.example.com:443
    facets:
    - https://proxy.example.com:443
    - https://proxy.example.com
    - proxy.example.com:443
    - proxy.example.com
```

If cloud wanted to limit the possible values of resources anyone (including admin) could create, cloud would define the following `constraints` companion type resource:

```yaml
kind: constraints
version: v2
spec:
  # sets constraints on all resources
  - kind: cluster_auth_preference
    fields:
        # only allow u2f or totp
        type: ['^u2f|totp']
```

Now, the constraints resource will follow the same resource flow. Cloud will hardcode the constraints in the config file, so no admin or user would be able to override it. What do you think?

klizhentas commented on rfd/0028-cluster-config-resources.md:
```diff
@@ -0,0 +1,298 @@
+kind: audit_config
+[...]
+---
+kind: cluster_auth_preference
```
is that already a new resource? the name `cluster_auth_config` will work better as it will match every other one

alex-kovoy commented:
> had similar thoughts but from the side of u2f logins. We don't want to allow admins to turn off second factor for the cloud - too risky. How do we achieve that?

@klizhentas can we do that in the code using feature flags? My thinking is that we are about to make it harder to use RBAC for the users because of the `cloud` implementation details.

klizhentas commented:
@alex-kovoy Constraints help to solve other problems - for example, admins can prohibit creating roles with agent forwarding enabled, or set max session TTL, so it enables more compex use cases.

Feature flags would have to have as many permutations as there are possible tweaks for cloud, this is what I would like to avoid.

andrejtokarcik commented:
@klizhentas The general `constraints` idea looks good to me.

Do I understand correctly that such constraints could be defined for any resource whatsoever, i.e. not only cluster-config-related resources? Are there any use cases that could be negatively impacted performance-wise by the constraint evaluation occurring during every resource creation/update? 

BTW, the assumption:
> Cloud will hardcode the constraints in the config file, so no admin or user would be able to override it.

cannot be taken for granted in an absolute sense. Even config-file specifications may happen to get overridden (even if temporarily) using `tctl create --force --confirm` (the underlying APIs should however require the `create` privilege in addition to mere `update`.)

klizhentas commented:
@andrejtokarcik I think the constraints performance hit would be negligible because you can use our existing cache logic with it. Constraints will be rarely upgraded/deleted, so 99.9% of the time they will be read from memory of the auth server that caches all resources all the time.

>  Even config-file specifications may happen to get overridden (even if temporarily) using tctl create --force --confirm (the underlying APIs should however in that case require the create privilege in addition to mere update).

We should have a mode when config file can not be overwritten, otherwise cloud integrity will be not possible. How do we do that?

andrejtokarcik commented:
> We should have a mode when config file can not be overwritten

Such a mode could be achieved by denying the `create` privilege for dynamically configurable resources, cf. #6463.

andrejtokarcik commented:
After some experimenting it seems there would be some serious hurdles involved in implementing the `constraints` resource in its full generality.

A generic proto definition for the resource I came up with is the following:
```proto
message ConstraintsV2 {
    // usual resource metadata stuff...

    repeated ResourceConstraintsSpecV2 Spec = 5
        [ (gogoproto.nullable) = false, (gogoproto.jsontag) = "spec" ];
}

message ResourceConstraintsSpecV2 {
    string kind = 1;
    FieldConstraintsMap fields = 2;
}

message FieldConstraintsMap { map<string, FieldConstraints> map = 1; }

message FieldConstraints { repeated FieldConstraint constraints = 1; }

message FieldConstraint {
    oneof constraint {
        FieldConstraintsMap subfields = 1;
        string pattern = 2;
        double minimum = 3;
        double maximum = 4;
    }
}
```

(This design is inspired by [JSON validation keywords](https://json-schema.org/draft/2020-12/json-schema-validation.html#rfc.section.6) -- it'd be reasonable for the constraints to be translated & embedded into the JSON validation schemas we already use, instead of having to add validation to each and every resource field manually or introducing custom reflection machinery for structure traversal.)

The YAML structure corresponding to this proto definition is quite cumbersome. The following would correspond to the example considered in https://github.com/gravitational/teleport/pull/6472#issuecomment-824229066: 

```
kind: constraints
version: v2
spec:
- kind: cluster_auth_preference
  fields:
    map:
      type:
        constraints:
        - Constraint:
            pattern: '^u2f|totp$'
      u2f:
        constraints:
        - Constraint:
            subfields:
              map:
                facets:
                  constraints:
                  - Constraint:
                      pattern: '^https://acme.teleport.sh$'
- kind: cluster_networking_config
  fields:
    map:
      type:
        constraints:
        - Constraint:
            minimum: 10
        - Constraint:
            maximum: 100
```

awly commented:
@alex-kovoy @klizhentas I think we should keep our RBAC design consistent and introduce new resource names + verbs. While it adds more RBAC "surface area", it's in line with all other API endpoints. Feature flags for Cloud would make reasoning about authz harder.

IIUC, the main reason for custom verbs or `constraints` is to restrict `SessionRecordingConfig` changes (i.e. switching to proxy recording mode).
How about we don't introduce anything complex right away, and just don't grant Cloud users permissions over that resource? That is, use new resource names (like `cluster_networking_config`) with standard verbs (create/list/read/write/delete).

The `constraints` idea is interesting, but also complex. It warrants a standalone design, which can be done after this RFD is implemented.

andrejtokarcik committed:
Let ClusterID become a ClusterName field

andrejtokarcik committed:
Remove section on data migration

No data migration should be needed since all (or almost all: excepting `ClusterID`) fields of the `ClusterConfig` resource are always loaded from the config file. The new code would simply use the same config file values to set a different resource.

klizhentas commented on rfd/0028-cluster-config-resources.md:
````diff
@@ -0,0 +1,288 @@
+be assigned a role featuring `update.no_proxy_recording` instead of the regular, more
+permissive `update`:
+```
+  allow:
````
I think it's an interesting concept, however this won't work unfortunately for cloud. For cloud you'd need to have some fields that can not be changed even by cluster administrator.

I propose we remove this change then out of scope of the RFD and instead use lib/modules to restrict some values that we know can not be changed by the cloud:

https://github.com/gravitational/teleport/blob/master/lib/modules/modules.go#L66

```
interface Modules {
   ValidateResource(res Resource) error
}
```

Then cloud build would implement 

```go
ValidateResource(res Resource) error {
   switch r := res.(type) {
        case *ClusterNetworkingConfig:
            if features.Cloud && r.GetMetadata().GetLabels().Source == "user" // whatever we defined for user {
                  return trace.BadParameter("cloud users not modify the networking parameters")
            }
       case *AuthConfig:
           if r.GetSecondFactor() == "off" {
               return trace.BadParameter("cloud can not have second factor off")
           }
}
}
```

andrejtokarcik replied:
Would it be OK to let this RFD deal only with the `ClusterConfig` reorganization and entirely omit this discussion of "Restricting to a subset of values of a field"?

awly replied:
@andrejtokarcik let's just add a short section explaining _how_ Cloud will restrict modifications (hardcoded checks via https://github.com/gravitational/teleport/blob/master/lib/modules/modules.go#L64-L75) without describing any specific limitations (e.g. "can't turn of 2FA").

We can let the Cloud team figure out what they want to restrict outside of this RFD.

andrejtokarcik replied:
Updated.

andrejtokarcik committed:
Update 0028-cluster-config-resources.md

awly commented:
@andrejtokarcik we'll survey users about the usage of `SetClusterConfig` API.
For now, let's continue with the assumption that it's not used outside of the Auth server itself.
Mention this in the RFD and let's not add any clever resource syncing code.

Note: `GetClusterConfig` should still return the merged config with all legacy fields.

andrejtokarcik commented:
Another question: Currently the RFD proposes to move the `ClusterID` field to the `ClusterName` resource. Even if not directly supported by `tctl`, the latter resource can be still fetched/updated via the Auth server API.

An older comment at https://github.com/gravitational/teleport/pull/5160#issuecomment-755031789 indicates that `ClusterID` shouldn't be exposed via server API at all. Wouldn't it be therefore preferable for `ClusterID` to be moved into its own resource?

awly commented:
@andrejtokarcik `ClusterName` shouldn't be changed either AFAIK.
Can that resource just be immutable via the API?

andrejtokarcik commented:
@awly For `ClusterName` we already have a set endpoint exposed: https://github.com/gravitational/teleport/blob/17ff928381025dc729ce1fa058004bbd75e2b6e5/lib/auth/apiserver.go#L204

So the situation with `ClusterName` is essentially the same as with the current `ClusterConfig`.

awly commented:
@andrejtokarcik I think it's in the same boat as the `SetClusterConfig` - the endpoint is there but nobody uses it outside of tests.
Let's keep the endpoint around just in case, but make it fail when `ClusterID` is set in the request to anything other than the current value.

andrejtokarcik commented:
@awly As mentioned in the linked older comment, the `ClusterID` field isn't really supposed to be even *viewable*. (Given that the `ClusterID` field is used as the secret key for HMAC anonymizing it might constitute a security issue as well.)

If that is the case, adding it to a resource that would otherwise be user-viewable/configurable (as `ClusterName` is probably meant to be) would require ad hoc gimmicks in the get handler as well. Therefore, my original question was about whether there actually should not be a dedicated resource for holding such private data.

andrejtokarcik committed:
Add a section on backward compatibility

andrejtokarcik committed:
Merge branch 'master' into rfd/0028-cluster-config-resources
